### DISCUSSION TITLE
Possible to use Deferrable feature alongside with TaskFlow?

### DISCUSSION THREAD

**Discussion by huxuan** [apache/airflow#20606] [General]
TaskFlow API [1] and Deferrable Operators [2] are both great features. Just want to know whether it is possible to combine them together. Thanks in advance.

[1] https://airflow.apache.org/docs/apache-airflow/stable/tutorial_taskflow_api.html
[2] https://airflow.apache.org/docs/apache-airflow/stable/concepts/deferring.html

**Comment by potiuk** [comment]
yes

**Reply by huxuan** [reply]
Any samples or tutorials on that? I feel a little confused about how the triggers and callback function work.

**Reply by potiuk** [reply]
Just the usual tutorial:

TaskFlow: https://airflow.apache.org/docs/apache-airflow/stable/concepts/taskflow.html
Defferable https://airflow.apache.org/docs/apache-airflow/stable/concepts/deferring.html?highlight=deferrable

There is nothing special about using deffereble feature in taskflow operators I believe. 

But if you go throught both tutorials and manage to make it working and feel that some special tutorial is needed for the combination - i heartily invite you to become one of the almost 1900 contributors and add such chapters/tutorial - it's super easy. Just use "suggest a change on this page" button at the bottom.

**Reply by potiuk** [reply]
Actually people who are confused and find their way are the best to write such tutorial, so I'd love to get one from you once you go through it.

**Reply by potiuk** [reply]
Basically, I believe you just need to throw the TaskDeferred exception from your decorated callable and it needs to be able to handle resuming by handling the parameters passed to the callable:

| Under the hood, self.defer raises the TaskDeferred exception, so it will work anywhere inside your Operator’s code, even buried many nested calls deep inside execute(). You are free to raise TaskDeferred manually if you wish; it takes the same arguments as self.defer.

**Reply by potiuk** [reply]
The only `problematic` thing is I believe that you can't have any other methods called - only the original callable method when resuming.

**Reply by huxuan** [reply]
@potiuk Thanks for the detailed explanation (and sorry for the late reply due to vocation). I will try it later. Will come up later then.

**Comment by Gollum999** [comment]
I've found a few ways of doing this, but they all seem quite hacky.

The main problem is `TaskDeferred.method_name`.  The callback needs to be a member of the `Operator` class, but the TaskFlow API doesn't really give you a natural opportunity to extend the class.

My attempts:

1. Monkey-patch `_PythonDecoratedOperator`.  Seems fragile, not sure if it would even work with a distributed deployment (I'm just testing with `airflow standalone`).
```
_PythonDecoratedOperator.defer_callback = lambda self, context, event: logger.info('in defer_callback')

@task
def test_defer1():
    raise TaskDeferred(
        trigger=TimeDeltaTrigger(delta=datetime.timedelta(seconds=10)),
        method_name='defer_callback',
    )

test_defer1()
```

2. Abuse an existing method with a compatible signature.  I'm sure I don't need to elaborate on why this is awful.
```
@task(post_execute=lambda context, event: logger.info('in post_execute'))
def test_defer2():
    raise TaskDeferred(
        trigger=TimeDeltaTrigger(delta=datetime.timedelta(seconds=10)),
        method_name='_post_execute_hook',
    )

test_defer2()
```

3. Custom operator wrapper.  Arguably the best option here, but very boilerplate-y.  IMO, needing to define a custom operator type defeats most of the purpose of the TaskFlow API.
```
class _PythonDecoratedOperatorWrapper(_PythonDecoratedOperator):
    def callback(self, context, event):
        logger.info('in wrapper callback')

def task_wrapper(python_callable=None, **kwargs) -> 'TaskDecorator':
    return task_decorator_factory(
        python_callable=python_callable,
        decorated_operator_class=_PythonDecoratedOperatorWrapper,
        **kwargs,
    )

@task_wrapper
def test_defer3():
    raise TaskDeferred(
        trigger=TimeDeltaTrigger(delta=datetime.timedelta(seconds=10)),
        method_name='callback',
    )

test_defer3()
```

Has anyone found a better way of doing this?

**Reply by potiuk** [reply]
I think this needs far more thoughts (and any of the solutions there are hacky). The main benefit of @task and taskflow is far less boilerplate. Anything that makes it more complex and adds boilerplate, and basically cancels out the benefit of using the task flow operators in the first place.

I think rather than monkeypatching this shoudl be a completely different approach. Since pyhon has a lot of functional aspects, so it should rather look like this (some details to be worked out):

```
    @task()
    def a_task(order_data_dict: dict):

        def complete():
            # do something
            pass

        # do_something before deferring
        raise TaskDeferred(trigger=TimeDeltaTrigger(timedelta(hours=1)), method_name="complete")
```

I have not looked though at some details but well, there should be no reason why it should not work IMHO (if we implement retrieving the current method context and find a way how to execute nested function. There are proabbly some consequences that I have not thought about as it needs a bit of python internal magic, and there is a question on how we should store/restore state of such function when resumed (or maybe make sure we do not have to have a state). But in principle, from the user's perspective that would be the nicest approach IMHO.  @andrewgodwin  - WDYT? - do you think it would be easy/doable to do ?

**Reply by safronau** [reply]
Any news on this topic? TaskFlow becames more and more popular, but what to do with deferrable is not clear still.

**Reply by potiuk** [reply]
I think we are waiting for someone to take a lead on that. As everything in open-source project, each change needs a lead. Maybe you would like to lead it @safronau if no-one else is ? Writing an AIP, discussing on devlist and implementing is what is needed.

**Comment by uranusjr** [comment]
Is there a more concrete use case for doing deferrals in a taskflow task? I think it would be better if we can understand the motivation behind it and work directly on improving the thing, instead of trying to mold either taskflow or deferrable into an abstract request. Maybe the solution is not taskflow + deferrable in the first place, but we should improve how a (class-based) deferrable task can be implemented, or something else entirely. All is only possible to unsderstand if there is a more concrete use case.

**Reply by VladimirYushkevich** [reply]
I have some "concrete" use case. Within a refactoring of dags and tasks in favour of Taskflow API  I found that it is not possible to use `@task` or `@sensor_task` annotation for the task that starts deferred trigger. This task is just a sensor very similar to mentioned in this [tutorial](https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/deferring.html#triggering-deferral). This looks to me that `@sensor_task` works not not for all sensors (at least without mentioned [here](https://github.com/apache/airflow/discussions/20606#discussioncomment-3181952) changes)

**Reply by uranusjr** [reply]
Hmm, `@task.sensor` (and `sensor_task`, they are the same thing) is intended to be a replacement to `PythonSensor`, not `BaseSensorOperator` implementations in general. We could re-brand it, sure, but that would be a more involved shift from simply “making” taskflow support deferrable.

What is the intention behind rewriting the sensor class into taskflow? Is it because the sensor is used in only one DAG (i.e. it should have used a `PythonSensor` in the first place) and you see this as a chance to simplify the code?

**Reply by VladimirYushkevich** [reply]
Right in our sensor we inherit from `BaseSensorOperator`. The intention to me is kind of simplification, yes. I mean if we can replace explicit definition of `PythonOperator` by `@task` marker on callable why is not possible to do the same for sensor?
Maybe I'm missing something about taskflow but just thought some new classes will support taskflow api.

**Reply by JarroVGIT** [reply]
One use case that we are having is that we are migrating to Airflow, and are not going to migrate everything straight away from our current scheduling tool. That means that we need to be able to start jobs on remote services (API calls) and wait for those jobs to be finished (whatever those jobs or services may entail doesn't really matter). By using deferrable operators, we were hoping to have light weight (non-slot taking) tasks so our DAGs would halt until they could be resumed again. 

We started off with `@task` based DAGs, but now it seems that this particular thing (wait for external system) must be implemented in another (~old~ classic) way. 

Is this not a typical use case of using the deferral feature?

**Reply by potiuk** [reply]
Classic != old. There are use-cases for both approaches. 

For example if you want to tap into the 3500+ operators that are implemented to handle 90+ services out there, using classic operator is best, for example we have really nice SQL operators that handle a lot of functionality, and so on. And there is no particular expectations that tasks are implemented (even in the same DAG) using exclusively one or the other approach. They can be mixed freely and our docs show how you can do it.

I understand your desire for consistency, but while we were discusing how to implement more "task-like" version of deferrable operator, we have not yet came to conclusion how it should be done., so until it's done you should use what's there. Maybe at some point of time all the use cases we have now for classic operators will find their "decorated" counterpart, but this is not even planned now.

BTW. If you have an idea how to do it better and propose a syntax and edge-case handling on how to make defferability with taskflow, you can.  Same like at least 2800 people who contributed ideas, code and documentation to Airflow - contribute your ideas by starting discussion on devlist and propising your idea for that. This is absolutely possible and welcome - you will find all the links at https://airflow.apache.org/community/

**Reply by JarroVGIT** [reply]
Thanks for the swift reply! I corrected my comment, I didn't mean 'old' in a condescending way toward the class based model and I apologize if that came across that way. 

On the point of contribution, I just cloned the repo to get more familiar with the inner workings of Airflow. If I come up with something, I will see if it meets the standards to warrant a PR :)

**Reply by RNHTTR** [reply]
> Is there a more concrete use case for doing deferrals in a taskflow task?

Another use case similar to @VladimirYushkevich 's idea is making deferrables more accessible to new Airflow users who prefer to develop DAGs strictly in TaskFlow. I don't think it makes sense to force users to leverage the operator paradigm, because that has an additional learning curve.
